mapper/cumulocity/c8y_smartrest/src/smartrest_deserializer.rs

```diff
@@ -164,6 +164,68 @@ impl SmartRestUpdateSoftwareModule {
     }
 }
 
+fn to_datetime<'de, D>(deserializer: D) -> Result<NaiveDateTime, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    // NOTE `NaiveDateTime` is used here because c8y uses for log requests a date time string which
+    // does not exactly equal `chrono::DateTime::parse_from_rfc3339`
+    // c8y result:
+    // 2021-10-23T19:03:26+0100
+    // rfc3339 expected:
+    // 2021-10-23T19:03:26+01:00
+    let date_string: String = Deserialize::deserialize(deserializer)?;
+    Ok(NaiveDateTime::parse_from_str(&date_string, "%Y-%m-%dT%H:%M:%S%z").unwrap())
+}
+
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+pub struct SmartRestLogRequest {
+    pub message_id: String,
+    pub device: String,
+    pub log_type: String,
+    #[serde(deserialize_with = "to_datetime")]
+    pub date_from: NaiveDateTime,
+    #[serde(deserialize_with = "to_datetime")]
+    pub date_to: NaiveDateTime,
+    pub needle: Option<String>,
+    pub lines: usize,
+}
+
+impl SmartRestLogRequest {
+    pub fn new() -> Self {
+        Self {
+            message_id: "522".to_string(),
+            device: "".to_string(),
+            log_type: "".to_string(),
+            date_from: NaiveDateTime::from_timestamp(0, 0),
+            date_to: NaiveDateTime::from_timestamp(0, 0),
+            needle: None,
+            lines: 0,
+        }
+    }
+
+    pub fn from_smartrest(smartrest: &str) -> Result<Self, SmartRestDeserializerError> {
+        let mut rdr = ReaderBuilder::new()
+            .has_headers(false)
+            .flexible(true)
+            .from_reader(smartrest.as_bytes());
+
+        let mut record: Self = Self::new();
+
+        for result in rdr.deserialize() {
+            record = result?;
+        }
+
+        Ok(record)
```

Comment:
Rather than mutating a default record, I would write:

```suggestion
        match rdr.deserialize().next() {
            Some(Ok(record)) => Ok(record),
            Some(Err(err)) => Err(err)?,
            None => panic!("empty request"),
        }
```

Rather than handling properly the empty  request case, I would make sure it cannot be. This can be done at a higher level that first extract the operation id then invoke this parse function once sure to have a log request. This means the match cases written in the `process_smartrest` function should be moved here in a function that return an enum (either a log request or ....).